design-proposals/klip-7.md

```diff
@@ -0,0 +1,176 @@
+
+In today's KSQL, the key for AGG is computed by concatenating the string forms of X and Y. In the future we may wish to create a structured key instead. However, we need to be sure to continue using the concatenated string format for queries that were started before that change was made.
+
+Our current approach to this problem is to do our best not to break compatibility by toggling incompatible behavior on compatibility-breaking configs, and to test that queries are built compatibly across versions. However, there are a number of problems with this approach:
+
+- The tests have some gaps. For example, we can't always catch incompatibility from pushing up filters or reordering internal projections
+- Tests are always incomplete, and breaking compatibility can lead to errors that are unrecoverable.
+- The tests rely on an unstructured topology format taken from Kafka Streams. This is very brittle. Part of this proposal is to switch to a more structured format.
+- The tests take a long time to run. Currently there are 473 test cases that must be repeated for every previous version of KSQL. This is going to grow as we support cloud, where we want to be able to release more frequently.
+- The usage of compatibility-breaking configs adds a lot of complexity to the codebase.
+
+In this one-pager, we propose building query execution plans at the SchemaKStream level that capture the low-level details of the computation being done at each step of a query. This plan can then be persisted and verified by compatibility-verification tests. Longer term, I propose that we persist this plan to the command topic (instead of statements + config), and build queries from the persisted execution plans instead of rebuilding the query from scratch each time.
+
+## What is in scope
+
+- Generate structured execution plan that describes all streams calls along with custom computations we do for KSQL.
+- Persist/verify the execution plan as part of query-validation tests.
+- Persist the execution plan to the command topic and build queries from it.
+
+## What is not in scope
+
+- Exposing the execution plan to users. I think this is worthwhile, and a clear improvement over the string that's returned today, but it's not what we're trying to solve here.
+
+## Value/Return
+
+- Less risk during upgrades.
+- Easier to maintain code.
+
+## Design
+
+First let's try to describe what it means for topologies to be compatible:
+
+- The two topologies should have the same set of sub-topologies.
+- The sub-topologies must be compatible:
+- The sub-topologies must have the same source Kafka topics (sources and state stores)
+- The sub-topologies must perform the same computation on the data read from the source Kafka topics
+- The sub-topologies must write the results out to the same Kafka topics (output and state stores) with the same formats.
+
+In the following sections I'll describe short term and longer term solutions to ensuring compatibility.
+
+### Short Term
```

Comment:
Controversial, but I think we should forgo the short-term and focus on the long term suggestion that you lay out (i.e. we should not write the code that compares two execution plans). I have a feeling that the code which compares if two serialized plans are compatible will be rather difficult, perhaps more difficult than re-creating the KStreams application from a single serialized plan. Also if design tradeoffs come along that make the implementation favor one or the other, I don't think we should make sacrifices for the short-term solution since the ROI of the long-term is so high (in my opinion).

Reply:
I should add a section on how we could stage this. Basically I was trying to draw a line in the sand between the minimal thing we want for compatibility testing as part of the ci/cd improvements vs the overall proposal.

It should actually not be too much of a detour to integrate this into the existing tests first and then change our execution model to leverage it:

1. As a first step, we can have SchemaKStream/SchemaKTable build the execution plan
2. The next step would be to persist execution plans in place of streams topologies in QueryTranslationTest. These can be checked for strict equality when verifying compatibility (which is why I think it won't actually be such a hard check to implement).
3. The next step would be to actually build the Kafka Streams app from the execution plans.
4. The last step would be to actually change the execution model to persist the execution plan to the command topic/config topic and build the final topologies from there.

Out of the above the only throwaway work would be step 2, which maybe is not too bad.